dask/array/utils.py

```diff
@@ -89,8 +101,10 @@ def assert_eq(a, b, check_shape=True, **kwargs):
             assert_eq_shape(b_original.shape, b.shape, check_nan=False)
     else:
         bdt = getattr(b, 'dtype', None)
+        if not hasattr(b, 'dtype'):
+            b = np.array(b)
 
-    if str(adt) != str(bdt):
+    if check_dtype and str(adt) != str(bdt):
```

Comment:
I'd remove this flag entirely, in favor of better handling of objects in `assert_eq`. If a result doesn't have a dtype attribute, it's implicitly an object dtype. We should always be checking the dtype of the results.